formfactory/tests/test_base.py

```diff
@@ -0,0 +1,269 @@
+        self.assertIn(self.validator, validators.get_registered_validators())
+
+        validators.unregister(self.validator)
+
+        # ensure the class is unregistered as expected
+        self.assertNotIn(
+            self.validator, validators.get_registered_validators()
+        )
+
+    def test_validation(self):
+
+        # ensure an excpetion is raised if the validation class is not complete
+        self.assertRaises(
+            NotImplementedError, self.incomplete_validator.validate, None
+        )
+
+        # ensure that the validate method returns correctly
+        self.assertTrue(self.validator.validate(4))
+
+
+class ActionTestCase(TestCase):
+    def setUp(self):
+        self.action = TestAction()
+        self.incomplete_action = TestActionIncomplete()
+
+    def test_registry(self):
+        actions.register(self.action)
+
+        # ensure the class is registered as expected
+        self.assertIn(self.action, actions.get_registered_actions())
+
+    def test_unregistry(self):
+        actions.register(self.action)
+
+        # ensure the class is registered as expected
+        self.assertIn(self.action, actions.get_registered_actions())
+
+        actions.unregister(self.action)
+
+        # ensure the class is unregistered as expected
+        self.assertNotIn(self.action, actions.get_registered_actions())
+
+    def test_action(self):
+
+        # ensure an excpetion is raised if the validation class is not complete
+        self.assertRaises(
+            NotImplementedError, self.incomplete_action.run, None
+        )
+
+        # ensure that the run method returns correctly
+        self.assertTrue(self.action.run({}))
+
+
+class ModelTestCase(TestCase):
+    def setUp(self):
+        load_fixtures(self)
+
+    def test_field_constant(self):
+
+        # ensure the field types are populated
+        self.assertIn(("DateTimeField", "DateTimeField"), models.FIELD_TYPES)
+        self.assertIn(("BooleanField", "BooleanField"), models.FIELD_TYPES)
+        self.assertIn(("CharField", "CharField"), models.FIELD_TYPES)
+
+    def test_form(self):
+
+        # ensure the form model has been saved correctly
+        for key, value in self.form_data.items():
+            self.assertEqual(getattr(self.form, key), value)
+
+        # ensure all form fields were saved
+        self.assertEqual(self.form.fields.count(), len(models.FIELD_TYPES))
+
+        # ensure the form object is returned
+        self.assertIsInstance(self.form.as_form(), forms.Form)
+
+    def test_formfield(self):
+
+        # ensure the form model has been saved correctly
+        for count in range(len(models.FIELD_TYPES)):
+            formfield_data = getattr(self, "formfield_data_%s" % count)
+            for key, value in formfield_data.items():
+                formfield = getattr(self, "formfield_%s" % count)
+                self.assertEqual(getattr(formfield, key), value)
+
+
+class AdminTestCase(TestCase):
```

Comment:
Split the tests up please. So crate eg. `test_admin.py`, `test_models.py` etc. Makes for easier reading.